### PR TITLE
Add hot reload apply changes API: AssemblyExtensions.ApplyUpdate

### PR DESCRIPTION
**Pull request by mikem8361**
Issue: https://github.com/dotnet/runtime/issues/45689

Currently Windows only: there will be another PR to enable this API on Linux/MacOS.

The Hot reload API will fail if debugging or if the module isn't editable.

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:70]: Messages should be in the .resx

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:87]: Is there any information we can include in the error message that would be more meaningful?

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:59]:
```suggestion
        /// <summary>
        /// Updates the specified assembly using the provided metadata, IL and PDB deltas.
        /// </summary>
        /// <remarks>
        /// Currently executing methods will continue to use the existing IL. New executions of modified methods will
        ///  use the new IL. Different runtimes may have different limitations on what kinds of changes are supported,
        /// and runtimes make no guarantees as to the state of the assembly and process if the delta includes
        /// unsupported changes.
        /// </remarks>
        /// <param name="assembly">The assembly to update.</param>
        /// <param name="metadataDelta">The metadata changes to be applied.</param>
        /// <param name="ilDelta">The IL changes to be applied.</param>
        /// <param name="pdbDelta">The PDB changes to be applied.</param>
        /// <exception cref="ArgumentNullException">The assembly argument is null.</exception>
        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:45]: Can we remove "HotReload" from the naming?

File: src/coreclr/vm/encee.cpp
Comment (stephentoub) [inline src/coreclr/vm/encee.cpp:400]:
```suggestion
    // Tell the debugger about the new method so it gets the version number properly
```

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs
Comment (stephentoub) [inline src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:13]: I'm surprised we decided to make the last parameter optional given that we don't actually expect many folks to be calling this method.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:60]: Is there a specification somewhere for what these deltas look like?

File: src/coreclr/vm/assemblynative.cpp
Comment (jkotas) [inline src/coreclr/vm/assemblynative.cpp:1440]: Why does this need to suspend the runtime? I think it should be possible to apply the delta without runtime suspended.

File: src/coreclr/vm/assemblynative.cpp
Comment (jkotas) [inline src/coreclr/vm/assemblynative.cpp:1438]: Why is this blocked when the debugger is attached? It should work just fine when the debugger is attached.

I understand that there is a problem with both debugger and this API apply conflicting deltas. We should detect that situation and only report error once that happens.

**Comment by jkotas** [comment]
Any tests to add?

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:70]: `throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);`

File: src/coreclr/vm/assemblynative.cpp
Comment (jkotas) [reply src/coreclr/vm/assemblynative.cpp:1438]: Also, blocking this API when the managed debugger is attached will make it really hard for engineers outside the core runtime team to debug any hot-reload related issues.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (mikem8361) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:70]: Done.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (mikem8361) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:59]: Done

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (mikem8361) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:87]: The runtime doesn't give us much info on failures.  The pinvoke returns an HRESULT from the vm code but there really isn't any detail there. It is either going to be E_FAIL, E_INVALIDARG or E_NOINTERFACE.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (mikem8361) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:45]: Done.

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs
Comment (mikem8361) [reply src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:13]: I'll remove the default on the pdbDelta parameter.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (mikem8361) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:60]: Not that I know of. @jkotas may have more info.

File: src/coreclr/vm/assemblynative.cpp
Comment (mikem8361) [reply src/coreclr/vm/assemblynative.cpp:1440]: Ok, I'll remove it. I add this in anticipation of the batching/multi-assembly changes discussed in the issue and never removed it.

**Comment by mikem8361** [comment]
We (Alexksey and I) are going to work on some tests. It will take a while to put them together and I didn't want to hold up getting in API so the dotnet-watch tooling (@pranavkm) can start using it.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:60]: I am not aware of any written spec for the delta format. @tmat do you about any?

The deltas are straightforward extension of the ECMA335. System.Reflection.Metadata has support to emit them (e.g. https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.ecma335.metadatabuilder.-ctor?view=net-5.0#System_Reflection_Metadata_Ecma335_MetadataBuilder__ctor_System_Int32_System_Int32_System_Int32_System_Int32_).

File: src/coreclr/vm/assemblynative.cpp
Comment (mikem8361) [reply src/coreclr/vm/assemblynative.cpp:1438]: I wanted to keep things simple and the current thoughts on the overall user experience does need for debugging and hot reload to work at the same time.  We want to (eventually) smoothly transition from hot reload changes to debugging (which is going to take work to update the debugger and DAC/DBI metadata copies). 

One of the biggest problems with allowing it is that debugger messages are sent to the right side when a function or field is added and that assumes the process is suspended/stopped in the debugger.

Comment (jkotas) [reply src/coreclr/vm/assemblynative.cpp:1438]: Can we just not let debugger know about any edits made using this API, for now at least?

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (noahfalk) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:89]: we could always change the runtime-side code to provide a more meaningful HRESULT too. 

From a priority standpoint I would guess we are expecting the Rosyln/dotnet-watch code to figure out when the edits aren't going to be successful prior to calling this method. However if there is some non-trivial slice of the scenario where dotnet-watch can't accurately predict success then the user is going to be dependent on the error information here to understand why things didn't work.

File: src/coreclr/vm/assemblynative.cpp
Comment (mikem8361) [reply src/coreclr/vm/assemblynative.cpp:1438]: I can add a "updateDebugger" flag to EditAndContinueModule::ApplyEditAndContinue and use that to control the debugger messages.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (tmat) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs]: I don't think we have a spec.

File: src/coreclr/vm/assemblynative.cpp
Comment (jkotas) [reply src/coreclr/vm/assemblynative.cpp:1438]: I think it would be useful to make the hot reload e2e easier to debug.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:89]: I agree that providing detailed exception for mall-formed delta is hard and we do not need to do it.

I think it would be useful to provide more details for operator errors, e.g. throw `InvalidOperationException` with appropriate message when somebody calls this on API on assembly that is not enabled for EnC.

Note that you can throw exception in the .cpp just fine. I would change the QCall to return void and throw all exceptions from the runtime side.

**Review by noahfalk — APPROVED** [review]
I anticipate we'll continue looking into the debugger interaction and error handling, but if the code as-is helps unblock anyone it seems like a fine starting point.

**Comment by noahfalk** [comment]
fyi @davidwrighton

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (mikem8361) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs]: Change made the change to throw from the cpp pinvoke.

File: src/coreclr/vm/assemblynative.cpp
Comment (jkotas) [inline src/coreclr/vm/assemblynative.cpp:1436]: Nit: We prefer to linear code where possible, e.g.:

```
if (CORDebuggerAttached())
    COMPlusThrow(...);

Module* pModule = assembly->GetDomainAssembly()->GetModule();
if (!pModule->IsEditAndContinueEnabled())
   COMPlusThrow(...);

if ((EditAndContinueModule*)pModule)->ApplyEditAndContinue(...))
     COMPlusThrow(...);
```

**Comment by jkotas** [comment]
> We (Aleksey and I) are going to work on some tests.

It should be straightforward to add basic tests that verify at least some of the error handling as part of this PR.

`src\libraries\System.Runtime.Loader\tests` is a good place to add tests like that.

File: src/coreclr/vm/assemblynative.cpp
Comment (mikem8361) [reply src/coreclr/vm/assemblynative.cpp]: I added some simple invalid parameter testing for the new API in the latest commit.

File: src/libraries/System.Runtime.Loader/tests/ApplyUpdateTests.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Loader/tests/ApplyUpdateTests.cs:1]: Nit: `AssemblyExtensionsTest.cs` may be a better name for this file. The convention is to use the same name for the file and the main type in the file.

**Review by jkotas — APPROVED** [review]
Thanks

File: src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs
Comment (lambdageek) [inline src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:32]: I'd weakly prefer if this was in a separate test - for now all assemblies are potentially editable in Mono (I'll revisit once we have a plan for https://github.com/dotnet/runtime/issues/47274).

It's ok if this is merged as is - I'll move it out when I rebase https://github.com/dotnet/runtime/pull/48380.

Comment (mikem8361) [reply src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:32]: You may want to consider not allowing System.Private.CoreLib.dll (which the AssemblyExtensions class is in) to be edited like coreclr does or what it considers "system" assemblies.

Comment (lambdageek) [reply src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:33]: >  You may want to consider not allowing System.Private.CoreLib.dll (which the AssemblyExtensions class is in) to be edited like coreclr does or what it considers "system" assemblies.

oh I went chasing down the wrong branch in `Module::IsEditAndContinueCapable` - thought this was the debug bits, not `PEFile::IsSystem`.  Yea it makes sense to not allow edits to CoreLib.

**Comment by mikem8361** [comment]
Anybody know off hand how to disable my new test for Mono runtimes?

**Pull request merged**
